server/map_tree_updater.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"sync"
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/extension"
+	"github.com/google/trillian/merkle/hashers"
+	"github.com/google/trillian/merkle/smt"
+	"github.com/google/trillian/storage"
+	"github.com/google/trillian/storage/tree"
+	stree "github.com/google/trillian/storage/tree"
+	"golang.org/x/sync/errgroup"
+)
+
+type mapTreeUpdater struct {
+	tree   *trillian.Tree
+	hasher hashers.MapHasher
+	opts   TrillianMapServerOptions
+	reg    extension.Registry
+}
+
+// update updates the sparse Merkle tree at the write revision of tx with the
+// given leaf changes, and writes it to the storage. Returns the new root hash.
+func (t *mapTreeUpdater) update(ctx context.Context, tx storage.MapTreeTX, upd []smt.NodeUpdate) ([]byte, error) {
```

Comment:
Wish list: the ability to be able to express a that these updates will be happening at an expected revision, and that no prior writes have occurred at this revision.

Reply:
Done to an extent: Now this accepts a `writeRev`, which is also verified to the bone in #1904.